extensions/amp-subscriptions/0.1/metering.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpDoc as AmpDocDef} from '../../../src/service/ampdoc-impl';
+import {Services} from '../../../src/services';
+import {dev} from '../../../src/log';
+
+/** @const */
+const TAG = 'amp-subscriptions';
+
+/** @const */
+const STORAGE_KEY = 'amp-subscriptions:metering-store';
+
+/**
+ * Describes metering state for a single publication.
+ * @typedef {{
+ *   id: string,
+ *   attributes: !Array<{
+ *     name: string,
+ *     timestamp: string,
+ *   }>,
+ * }}
+ */
+let MeteringStateDef;
+
+/** Handles metering functionality. */
+export class Metering {
+  /**
+   *
+   * @param {{
+   *   ampdoc: !AmpDocDef,
+   *   platformKey: string,
+   * }} params
+   */
+  constructor({ampdoc, platformKey}) {
+    /** @private {Promise<!../../../src/service/storage-impl.Storage>} */
+    this.storagePromise_ = Services.storageForDoc(ampdoc);
+
+    /**
+     * Remembers if metering entitlements were fetched
+     * with the current metering state.
+     *
+     * This helps avoid redundant fetches.
+     * @type {boolean}
+     */
+    this.entitlementsWereFetchedWithCurrentMeteringState = false;
+
+    /**
+     * Key for the subscription platform with metering enabled.
+     * @const
+     */
+    this.platformKey = platformKey;
+  }
+
+  /**
+   * Saves metering state for a given publication ID.
+   *
+   * @param {!MeteringStateDef} meteringState
+   * @return {!Promise}
+   */
+  saveMeteringState(meteringState) {
+    const meteringStateString = JSON.stringify(meteringState);
```

Comment:
Is JSON.stringify() consistent, will it always give the same output given the same object?  I seem to recall that the order the keys were added to the object is important.  In our use case this is probably ok if the publisher is consistent in how they order the metering state response but maybe add a debug user().info() if there are 2 strings and they are different ...

Reply:
Good call, it's now logging when an existing metering state gets updated